firedrake/slope_limiter/vertex_based_limiter.py

```diff
@@ -37,26 +37,30 @@ def __init__(self, space):
         self.centroid_solver = self._construct_centroid_solver()
 
         # Update min and max loop
-        self._min_max_loop = """
-for(int i = 0; i < maxq.dofs; i++) {
-    maxq[i][0] = fmax(maxq[i][0],q[0][0]);
-    minq[i][0] = fmin(minq[i][0],q[0][0]);
-}
-                             """
+        domain = "{[i]: 0 <= i < maxq.dofs}"
+        instructions = """
+        for i
+            maxq[i] = fmax(maxq[i], q[0])
+            minq[i] = fmin(minq[i], q[0])
+        end
+        """
+        self._min_max_loop = (domain, instructions)
+
         # Perform limiting loop
-        self._limit_kernel = """
-double alpha = 1.0;
-double qavg = qbar[0][0];
-for (int i=0; i < q.dofs; i++) {
-    if (q[i][0] > qavg)
-        alpha = fmin(alpha, fmin(1, (qmax[i][0] - qavg)/(q[i][0] - qavg)));
-    else if (q[i][0] < qavg)
-        alpha = fmin(alpha, fmin(1, (qavg - qmin[i][0])/(qavg - q[i][0])));
-}
-for (int i=0; i<q.dofs; i++) {
-    q[i][0] = qavg + alpha*(q[i][0] - qavg);
-}
-                             """
+        domain = "{[i, ii]: 0 <= i < q.dofs and 0 <= ii < q.dofs}"
+        instructions = """
+        <float64> alpha = 1
+        <float64> qavg = qbar[0, 0]
+        for i
+            <float64> _alpha1 = fmin(alpha, fmin(1, (qmax[i] - qavg)/(q[i] - qavg)))
+            <float64> _alpha2 = fmin(alpha, fmin(1, (qavg - qmin[i])/(qavg - q[i])))
+            alpha = if(q[i] > qavg, _alpha1, if(q[i] < qavg, _alpha2, alpha))
+        end
+        for ii
+            q[ii] = qavg + alpha * (q[ii] - qavg)
+        end
```

Comment:
Is the scheduling deterministic @inducer? i.e. if I compile this stuff in parallel, will all processes agree on the instruction order?

Reply:
Oops, is this the reason that this fails in parallel when vectorised?

Reply:
No idea, because I don't know what the failure mode is.

Reply:
So it fails when we both vectorise and run in parallel, but not when we do only one of the two. And this is the only failure with vectorisation, which is quite weird. I will try to get to the end of this.

Reply:
But what does "fail" mean. If the kernels are different on different processes you get a CompilationError. But it doesn't sound like you have that.

Reply:
Indeed.
So it looks the problem is with vectorising vertex based limiters in general. When we batch elements, e.g. batch = cells 0, 1, 2, 3, cell 1 does not see the updates of the vertices of cell 0 because the values are loaded in batches before the updates.
Is this algorithm intrinsically serial?
And I'm not sure why the test passes with np=1, probably just coincidence.

Reply:
Ah, ok. This is almost certainly because RW is a lie as an access descriptor. Instead it should be MIN or MAX. We will try and fix it this week.